Allow stylus support on windows

CodeDoctorDE opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

This is my attempt for adding support for stylus on windows.
~~**It's work in progress! I post it here to indicate that I'm working on it and that the same work is not done twice**~~

I'm working on the `flutter_window.cpp` file. Maybe I need to touch some other code to allow additional information like pressure, rotation and more.

Would fix https://github.com/flutter/flutter/issues/102836 and https://github.com/flutter/flutter/issues/65248

There should be no breaking change other than stylus would be reported as `stylus` instead of `touch`.

Tested with flutter master on my test project:
![image](https://github.com/user-attachments/assets/e8b75f96-2334-45e9-be23-753f1565a246)

https://github.com/CodeDoctorDE/flutter-input-demo

~~Tests are currently not there, I will add them next.~~ (Added).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

CodeDoctorDE committed:
Start replacing WM_TOUCH usage with WM_POINTER

CodeDoctorDE committed:
Fix button in pointer events

CodeDoctorDE committed:
Add check for incontact to remove hover recognized as move

CodeDoctorDE commented:
Looks like we can also identify hovering, I will work on this next

CodeDoctorDE commented:
Dont know why mac and linux ci fails, but my changes doesnt break any windows test so it looks good. Now i need to add my own tests

CodeDoctorDE committed:
Add pressure information to embedder and windows shell

Piinks commented:
Hey @CodeDoctorDE welcome! Thanks for contributing! It looks like this never entered the review queue since it was marked a draft. Is this still something you would like to work on?

CodeDoctorDE commented:
Hello, oh yeah, but it currently has some problems I need to fix

CodeDoctorDE commented:
Hello,
Yeah, Im happy to work again here, but in the current time, I hadn't enough time because of studying. I will continue here in a month

Piinks commented:
Quite alright! Thanks for letting us know. I am going to close this for now to remove it from our review queue and CI tasks. Feel free to reopen when you are ready to return to it. Thanks!

CodeDoctorDE commented:
Hello, I finished my changes but I don't have the option to reopen it.
Can you reopen it or should I make a new one?

CodeDoctorDE committed:
Finish pointer tests and win calls
by @Demohstens

Adds six unittests for different pointer events.
Added methods to WindowsProcTable for Pointer functions to allow for overloading in tests.
Ensured all unittest pass

Co-authored-by: Demohstens <Demohstens@gmail.com>

tirth-patel-nc commented:
> PR can be reviewed, not sure why mac fais

It is passing. Tree status would be green on its own

Demohstens committed:
Added kMaxPenPressure constant (#3)

Demohstens committed:
Added cmath dependancy to windows (#4)

* Added kMaxPenPressure constant

* Added cmath as dependancy to windows

loic-sharma commented on engine/src/flutter/shell/platform/windows/flutter_window.cc:
```diff
@@ -487,7 +492,6 @@ LRESULT CALLBACK FlutterWindow::WndProc(HWND const window,
     auto that = static_cast<FlutterWindow*>(cs->lpCreateParams);
     that->window_handle_ = window;
     that->text_input_manager_->SetWindowHandle(window);
-    RegisterTouchWindow(window, 0);
```
Ah gotcha. 

cc @jnschulze in case you have any concerns since you added this back in: https://github.com/flutter/engine/pull/27863

loic-sharma commented on engine/src/flutter/shell/platform/windows/flutter_windows_view.cc:
```diff
@@ -642,6 +651,13 @@ void FlutterWindowsView::SendPointerEventWithData(
   event.device = state->pointer_id;
   event.buttons = state->buttons;
   event.view_id = view_id_;
+  event.rotation = state->rotation;
```
Ah yup, looks like this is radians: https://api.flutter.dev/flutter/dart-ui/PointerData/rotation.html

That looks good to me, thanks!

loic-sharma commented:
I'll be on vacation for the next 10 days and won't be able to review. Apologies for the delay!

Demohstens committed:
Adressed comments by refactoring as well as adding (#5)

trackpad device type

loic-sharma commented:
@Demohstens please feel free to ping me when this is ready for another review :)

Demohstens commented:
Should be ready @loic-sharma - I totally forgot to mention you, sorry! 

If someone has a pen with rotation capabilities it'd be great to test that property. The tests don't account for the transformation as they intercept the function before it's sent to the framework and my pen does not support rotation :/

loic-sharma commented on engine/src/flutter/shell/platform/windows/flutter_windows_view.cc:
```diff
@@ -18,6 +19,11 @@
 namespace {
+// The windows API sends pressure as a normalized value between 0 and 1024
+// See
+// https://learn.microsoft.com/en-ca/windows/win32/api/winuser/ns-winuser-pointer_pen_info
```
The Canadian locale might not be the best one for the user, I'd let Microsoft's docs redirect :)

```suggestion
// https://learn.microsoft.com/windows/win32/api/winuser/ns-winuser-pointer_pen_info
```

CodeDoctorDE committed:
Address variable naming issues and docs link

CodeDoctorDE committed:
Rename x and y pos variables on all places inside handle message

mattkae reviewed (commented):
A few comments, but nothing huge on my side. I have no hardware to test though, so I'll have to trust 😄

CodeDoctorDE committed:
Apply review suggestions

CodeDoctorDE commented on engine/src/flutter/shell/platform/windows/flutter_window.cc:
```diff
@@ -537,47 +541,67 @@ FlutterWindow::HandleMessage(UINT const message,
+            device_kind = kFlutterPointerDeviceKindMouse;
+            break;
+          case PT_TOUCHPAD:
+            device_kind = kFlutterPointerDeviceKindTrackpad;
```
fixed

CodeDoctorDE commented on engine/src/flutter/shell/platform/windows/flutter_windows_view.cc:
```diff
@@ -644,6 +659,13 @@ void FlutterWindowsView::SendPointerEventWithData(
+    // Normalized between 0 and 1024 by the windows API
+    event.pressure_min = 0;
+    event.pressure_max = kMaxPenPressure;
+  }
```
fixed

mattkae commented on engine/src/flutter/shell/platform/windows/flutter_window.cc:
```diff
@@ -537,47 +541,71 @@ FlutterWindow::HandleMessage(UINT const message,
+            break;
+          default:
+            assert(false, 'Unrecognized device key {}', pointerInfo.pointerType);
+            return kFlutterPlane.value;
```
I don't think that we use `assert` in the code often? Could this be:

```suggestion
            FML_LOG(ERROR) << "Unrecognized device key " << pointerInfo.pointerType;
            break;
```

I would suggest `break`ing here just so that we don't "handle" a message that we don't know about. It can bubble down to Win32

CodeDoctorDE replied:
Sure, I'm not an expert in this codebase, I just looked how it is handled in other places. i commited the suggestion

mattkae reviewed (commented):
Nice tests 😄 I just have one final piece of feedback here

CodeDoctorDE committed:
Update engine/src/flutter/shell/platform/windows/flutter_window.cc

Co-authored-by: Matthew Kosarek <matt.kosarek@canonical.com>

mattkae reviewed (approved):
I am perfectly happy with this, thanks for putting in so much work on testing 🎉 

@loic-sharma Would you have anything to add? I cannot test on my end unfortunately

loic-sharma commented:
The tests that failed look like infra problems. I reran them!

CodeDoctorDE committed:
Fix formatting

CodeDoctorDE commented:
Oh yeah. I had a lot of stress the last months. I try to fix the problems in 2 weeks

CodeDoctorDE committed:
Fix compiling errors appearing after merging

loic-sharma reviewed (approved):
Thank you for the wonderful contribution! The code changes look good to me.

Unfortunately, I don't have the capacity to test this manually.

Piinks committed:
Merge branch 'master' into windows/pointer
